configuration/exodus-pipeline-resources.yaml

```diff
@@ -219,3 +229,44 @@ Resources:
       Artifacts:
         Type: NO_ARTIFACTS
       TimeoutInMinutes: 10
+
+  WhitelistUploadResource:
+    Type: Custom::WhitelistUploadFunction
+    Condition: UploadWhitelist
+    Properties:
+      ServiceToken: !GetAtt WhitelistUploadFunction.Arn
+      DestBucketName: !Sub ${project}-pipeline-artifacts
+      DestBucketRegion: !Ref region
+      WhitelistUrl: !Ref whitelistUrl
+    DependsOn:
+      - WhitelistUploadFunction
+
+  WhitelistUploadFunction:
+    Type: AWS::Lambda::Function
+    Condition: UploadWhitelist
+    Properties:
+      Handler: index.handler
+      Role: !Ref lambdafunctionrole
+      FunctionName: !Sub ${project}-whitelist-upload
+      Code:
+        ZipFile: |
+          import urllib.request
+          import cfnresponse
+          import boto3
+
+
+          def handler(event, context):
+              bucket = event["ResourceProperties"]["DestBucketName"]
+              region = event["ResourceProperties"]["DestBucketRegion"]
+              whitelist_url = event["ResourceProperties"]["WhitelistUrl"]
+
+              # Upload whitelist only when this resource is created
+              if event["RequestType"] == "Create":          
+                  s3client = boto3.client("s3", region_name=region)
+
+                  whitelist = urllib.request.urlopen(whitelist_url).read()
+                  s3client.put_object(Body=whitelist, Bucket=bucket, Key=".whitelist.gpg")
```

Comment:
@rohanpm 
So, I know you'd prefer to have the initial public keys hard-coded here but this implementation has file size limitations[1]. This means it'd be tough to fit more than a single key in this way. I think devs would have a hard time with this approach also, as they'd most likely need to hack this function for a smooth stack creation. Perhaps a compromise is to allow the user to pass in a public key string. 

[1] https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html#cfn-lambda-function-code-zipfile

CC: @crungehottman @dichen16

Reply:
I'm not liking this, overall it doesn't feel very clean.

- the python code embedded directly here means it's not covered by all the usual quality checks on our python code, e.g. the autotests and code formatter.
- it could presumably be tested & packaged like our other lambdas, but it seems cumbersome.
- the lambda function is sitting there all the time although it's never used after initial bootstrapping
- is the lambda missing some error handling?  What happens if the URL is not fetchable - we never send any cfnresponse in that case, so does it just time out?

I have two alternative suggestions to consider:

1. Have this whitelist put in place by the ansible role rather than in this template.  It does mean that this template won't give you absolutely everything you need to have a working pipeline, which is a shame, but the code will end up so much cleaner it's probably worth it.  aws_s3 ansible module seems to provide everything we need, it would look something like:

```
aws_s3:
  bucket: "{{ exodus_artifacts_bucket }}"
  mode: put
  object: .whitelist.gpg
  # by default, only put whitelist if it's absent, but can force overwrite
  # by setting exodus_whitelist_overwrite=always
  overwrite: "{{ exodus_whitelist_overwrite|default('never') }}"
  # This would be pointed at a whitelist file checked into the playbooks repo
  src: "{{ exodus_whitelist }}"
  # (... and all the usual aws credentials etc attributes here)
```

One thing I like about this is: just as we initially deploy the pipeline by running a job in ansible tower, if the need arises to manually reset the whitelist, we just need to re-run the ansible job with a parameter asking it to reset.

2. How about, in the pipeline, if the whitelist file doesn't exist in s3 then we just don't verify?  It means the very first deployment isn't verified, which isn't ideal, but as a one-time event it seems acceptable (I think I mentioned that in discussions around the original issue).  It doesn't seem to reduce the security of regular deployments - if an attacker had permission to delete the whitelist file, then they also could have replaced it with a whitelist containing their key, so no real change there.

I don't quite like this as much as (1) since it would be nice to have verification right from the beginning, and also since the method of resetting the whitelist would seem to be "manually delete it from s3" which is worse than having an ansible job doing it.

Reply:
Option 1 sounds best but if we're going to store a whitelist in the Ansible role repo, we may as well make that the sole copy. By this I mean removing the whitelist from release-engineering/exodus-lambda and therefore the final command of the verify-source CodeBuild project, which refreshes .whitelist.gpg in the bucket upon verification.

The whitelist in the role repo would be vaulted to prevent hijacking and serve as our single source of truth for authorizing stage & prod exodus-lambda updates. This would mean that any update to the whitelist wouldn't take effect until the playbook is run again. In other words, a newly added developer wouldn't be able to successfully trigger the stage pipeline until the playbook is re-run (and stacks updated).

Thoughts?
@rohanpm 
@crungehottman

Reply:
Yeah, I guess we don't really need the automated updating of whitelist.gpg from this repo.  Maybe simpler is best, meaning the ansible solution.

Reply:
Alright. I'll close this PR and update the Jira issue.